### PR TITLE
cudaPackages.cudatoolkit: switch to autoPatchelf

### PR DESCRIPTION
**Pull request by SomeoneSerge**
###### Description of changes

Rewrites `cudatoolkit` expression to use `autoPatchelf` instead of manually constructing and writing the `rpath`.
Using `autoPatchelf` ensures that we're at least not missing dependencies that upstream has marked as "needed".

This is a narrow-scoped part of https://github.com/NixOS/nixpkgs/pull/178439
For instance, this PR ensures "correctness" (amend missing rpaths) but increases the actual closure size.
The next PR should split the output to reduce closure sizes, while preserving "correctness"

###### Things done

- Built on platform(s)
  - [x] x86_64-linux

CC @NixOS/cuda-maintainers

### CODE REVIEW

**Comment by SomeoneSerge** [comment]
```
  dontPatchELF = true;
  dontStrip = true;
```

These two are just hanging around. I'm almost certain I should remove `dontPatchELF`. I'm not so sure about `dontStrip`

File: pkgs/development/compilers/cudatoolkit/common.nix [hunk 1]
```diff
@@ -189,47 +262,16 @@ stdenv.mkDerivation rec {
...
-  # it, just patch all (but not the stubs libraries). Note that
-  # --force-rpath prevents changing RPATH (set above) to RUNPATH.
-  postFixup = ''
-    addOpenGLRunpath --force-rpath {$out,$lib}/lib/lib*.so
-  '' + lib.optionalString (lib.versionAtLeast version "11") ''
-    addOpenGLRunpath $out/cuda_sanitizer_api/compute-sanitizer/*
-    addOpenGLRunpath $out/cuda_sanitizer_api/compute-sanitizer/x86/*
-    addOpenGLRunpath $out/target-linux-x64/*
-  '';
+    in "";
```
Comment (samuela) [inline pkgs/development/compilers/cudatoolkit/common.nix:274]: What's going on here? Where is `rpath` being used?

File: pkgs/development/compilers/cudatoolkit/common.nix [hunk 2]
```diff
@@ -53,12 +62,76 @@ stdenv.mkDerivation rec {
...
+    wayland
+  ];
+
+  # Prepended to runpaths by autoPatchelf.
+  # The order inherited from older rpath preFixup code
   runtimeDependencies = [
-    ncurses5 expat python27 zlib glibc
-    xorg.libX11 xorg.libXext xorg.libXrender xorg.libXt xorg.libXtst xorg.libXi xorg.libXext
-    gtk2 glib fontconfig freetype unixODBC alsa-lib
+    (placeholder "lib")
```
Comment (samuela) [inline pkgs/development/compilers/cudatoolkit/common.nix:113]: What is the `placeholder` function doing? Where is it defined?

**Review by samuela — COMMENTED** [review]
Looks like a great upgrade to cudaPackages.cudatoolkit! I'm running nixpkgs-review now...

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 178440` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.caffeWithCuda</li>
    <li>truecrack-cuda</li>
  </ul>
</details>
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>caffeWithCuda</li>
    <li>ethminer (ethminer-cuda)</li>
    <li>gpu-screen-recorder</li>
    <li>gpu-screen-recorder-gtk</li>
    <li>python39Packages.caffeWithCuda</li>
  </ul>
</details>
<details>
  <summary>33 packages built:</summary>
  <ul>
    <li>colmapWithCuda</li>
    <li>cudaPackages.cuda-samples</li>
    <li>cudatoolkit (cudaPackages.cudatoolkit ,cudatoolkit_11)</li>
    <li>cudaPackages.cutensor</li>
    <li>cudaPackages.nccl</li>
    <li>forge</li>
    <li>gpu-burn</li>
    <li>gromacsCudaMpi</li>
    <li>gwe</li>
    <li>katagoWithCuda</li>
    <li>librealsenseWithCuda</li>
    <li>magma</li>
    <li>nvtop</li>
    <li>nvtop-nvidia</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.numbaWithCuda</li>
    <li>python310Packages.pycuda</li>
    <li>python310Packages.pynvml</li>
    <li>python310Packages.pyrealsense2WithCuda</li>
    <li>python310Packages.pytorchWithCuda</li>
    <li>python39Packages.TheanoWithCuda</li>
    <li>python39Packages.cupy</li>
    <li>python39Packages.jaxlibWithCuda</li>
    <li>python39Packages.numbaWithCuda</li>
    <li>python39Packages.pycuda</li>
    <li>python39Packages.pynvml</li>
    <li>python39Packages.pyrealsense2WithCuda</li>
    <li>python39Packages.pytorchWithCuda</li>
    <li>python39Packages.tensorflowWithCuda</li>
    <li>xgboostWithCuda</li>
    <li>xpraWithNvenc</li>
  </ul>
</details>

**Comment by samuela** [comment]
OTOH the failures are only in the `extras/demo_suite/*`folder. We could just remove those binaries or skip them.

**Comment by SomeoneSerge** [comment]
I re-based on current `master`, and ran

```bash
❯ nix-build with-my-cuda.nix -A cudaPackages_10.cudatoolkit.out -A cudaPackages_10_1.cudatoolkit.lib -A cudaPackages.cudatoolkit
/nix/store/80q92g4mw49cifxfzhk3xhfmhcq7635p-cudatoolkit-10.2.89
/nix/store/vfkbm851wjfpw1pc2mdxq5v7d49plkps-cudatoolkit-10.1.243-lib
/nix/store/chzf3k3s07wd9i7xgzg6ha667bjhpc51-cudatoolkit-11.7.0
```

...nixpkgs-review would be nice, but I probably can't run it any time soon

**Comment by SomeoneSerge** [comment]
I think it would be pragmatic to just merge, relying on autoPatchelf having verified all the declared dependencies. There may be hidden `dlopen()` errors, but more likely in tools, not in libraries used by our ML stack. We can address these errors as they appear

File: pkgs/development/compilers/cudatoolkit/common.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/cudatoolkit/common.nix:121]: That is a builtin function of derivation which returns what the path of that output will be in nix code. In bash you could just write ``$lib``.

**Comment by MrFoxPro** [comment]
Hello @SomeoneSerge. I recently updated nixpkgs channel confugration and xmrig-cuda library that depends on cuda stopped working with this error: `failed to open libnvrtc-builtins.so.11.7`

I fixed it by downgrading nixpkgs to `7018cf78c618e0a8ec4369c587319f51cb7b19b0` and nvidia-driver
You can see my derivation here: https://github.com/MrFoxPro/nix/blob/cuda-bug/drv/xmrig-cuda.nix
It builds fine, but fails at runtime

**Comment by SomeoneSerge** [comment]
@MrFoxPro Hey-hey, и Вам доброго дня!

First off, I see that you're linking to `nvidia_x11` directly, which is something we try to avoid in nixpkgs: we deploy `libcuda.so` at `/run/opengl-driver/lib`, because it's driver-dependend. You might want to replace that with `autoAddOpenGLRunpathHook`.

As for the libnvrtc error and whether it's not being found or being rejected by the dynamic linker, we'll need to see more logs. I'd start with running xmrig with `LD_DEBUG=libs` environment variable set. The error could indeed be related to this PR, because we're now setting Runpaths more consistently. The only regression we have noticed ourselves so far is one linked from the pytorch PR. Obviously, though, we can only really see how we affect packages that are in nixpkgs, and we can sometimes break things out-of-tree even if we're careful 🙃 

I also see that there is an `xmrig` derivation in nixpkgs, only without cuda support yet. Maybe you could open a PR adding CUDA support to that derivation, and we could navigate from there?

**Comment by MrFoxPro** [comment]
> @MrFoxPro Hey-hey, и Вам доброго дня!
> 
> First off, I see that you're linking to `nvidia_x11` directly, which is something we try to avoid in nixpkgs: we deploy `libcuda.so` at `/run/opengl-driver/lib`, because it's driver-dependend. You might want to replace that with `autoAddOpenGLRunpathHook`.
> 
> As for the libnvrtc error and whether it's not being found or being rejected by the dynamic linker, we'll need to see more logs. I'd start with running xmrig with `LD_DEBUG=libs` environment variable set. The error could indeed be related to this PR, because we're now setting Runpaths more consistently. The only regression we have noticed ourselves so far is one linked from the pytorch PR. Obviously, though, we can only really see how we affect packages that are in nixpkgs, and we can sometimes break things out-of-tree even if we're careful 🙃
> 
> I also see that there is an `xmrig` derivation in nixpkgs, only without cuda support yet. Maybe you could open a PR adding CUDA support to that derivation, and we could navigate from there?

I'm not sure about `/run/opengl-driver/lib`. Does this exist only when `hardware.opengl.enable`is true? I mean I'm running miner on my machine headlessly and starting fake xserver only for overclocking via nvidia-settings, so I'm not sure why this option should be mandatory then.

**Comment by MrFoxPro** [comment]
> @MrFoxPro On a side note, though... I don't know what you mean by the "PMC Balloon", but to me it sounds a little bit provoking, and not in a good way. This is going off-topic though

just a meme :) you're welcome to join https://t.me/ru_nixos btw, so we can discuss it more closely if you want

**Comment by SomeoneSerge** [comment]
> I'm not sure about /run/opengl-driver/lib. Does this exist only when hardware.opengl.enableis true? I mean I'm running miner on my machine headlessly and starting fake xserver only for overclocking via nvidia-settings, so I'm not sure why this option should be mandatory then.

The name `hardware.opengl.enable` is historical legacy and subject to change: https://github.com/NixOS/nixpkgs/issues/141803. One is expected to use `hardware.opengl.enable` (and `videoDrivers = [ "nvidia" ]`, iirc) even in headless mode, so that programs from nixpkgs will know to use `libcuda.so` that is compatible with your system's driver. This does not imply enabling the X server. In fact, you don't necessarily need the X server even to use OpenGL (cf. EGL). For why we need to deploy `libcuda.so` impurely, cf. this comment: https://github.com/NixOS/nixpkgs/issues/224294#issuecomment-1494016185

But let's draft a PR and move the conversation there!

> just a meme :)

Alright, alright, I didn't mean to imply anything. Just that there are times when waving a white flag high above your head before approaching people is suddenly a very common sense thing to do, lest you catch friendly fire

**Comment by MrFoxPro** [comment]
@SomeoneSerge lets discuss it https://github.com/NixOS/nixpkgs/pull/224848

**Pull request merged**